src/builders/mmdf.jl

```diff
@@ -0,0 +1,70 @@
+
+# Copyright (c) 2021-2024, University of Luxembourg
+# Copyright (c) 2021-2024, Heinrich-Heine University Duesseldorf
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+$(TYPEDSIGNATURES)
+
+Allocate log-concentration-stoichiometry constraints for the `reaction_subset`
+and `metabolite_subset` in `model`, as used by e.g.
+[`max_min_driving_force_analysis`](@ref).
+
+The output constraint tree contains a log-concentration variable for each
+metabolite in subtree `log_concentrations`. The reaction quotient in log
+variables for each reaction is stored in `log_concentration_stoichiometry`.
+
+Function `concentration_bound` may return a bound for the log-concentration of a
+given metabolite (compatible with `ConstraintTrees.Bound`), or `nothing`.
+"""
+function log_concentration_constraints(
+    model::A.AbstractFBCModel;
+    reaction_subset = A.reactions(model),
+    metabolite_subset = A.metabolites(model),
+    concentration_bound = _ -> nothing,
+)
+    stoi = A.stoichiometry(model)
+
+    constraints =
+        :log_concentrations^C.variables(
+            keys = Symbol.(metabolite_subset),
+            bounds = concentration_bound.(metabolite_subset),
+        )
+
+    # map old idxs of metabolites to new order of smaller system
+    midx_lookup = Dict(
```

Comment:
kill indexes, we worked hard to finally eradicate the stupid indexing from everywhere.

this can be done by just using the original code and adding a conditional `continue` to the right spot